File: core/src/commonMain/kotlin/work/socialhub/kslack/entity/block/element/MultiStaticSelectElement.java
package work.socialhub.kslack.entity.block.element;

import com.github.seratch.jslack.api.model.block.composition.ConfirmationDialogObject;
import com.github.seratch.jslack.api.model.block.composition.OptionGroupObject;
import com.github.seratch.jslack.api.model.block.composition.OptionObject;
import com.github.seratch.jslack.api.model.block.composition.PlainTextObject;

import java.util.List;

/**
 * https://api.slack.com/reference/block-kit/block-elements#multi_select
 */
public class MultiStaticSelectElement extends BlockElement {
    public static final String TYPE = "multi_static_select";
    private final String type = TYPE;
    private String fallback;

    private PlainTextObject placeholder;
    private String actionId;

    // https://github.com/seratch/jslack/pull/103
    // The reason I didn't initialize the List<> fields is because Slack (sometimes) gives errors
    // when it encounters an empty list in the generated JSON.
    // The proper solution if/when you don't want un-initialized fields is to have a Gson type adapter that skips empty lists
    // (e.g. something like https://stackoverflow.com/questions/11942118/how-do-you-get-gson-to-omit-null-or-empty-objects-and-empty-arrays-and-lists)
    private List<OptionObject> options;

    // https://github.com/seratch/jslack/pull/103
    // The reason I didn't initialize the List<> fields is because Slack (sometimes) gives errors
    // when it encounters an empty list in the generated JSON.
    // The proper solution if/when you don't want un-initialized fields is to have a Gson type adapter that skips empty lists
    // (e.g. something like https://stackoverflow.com/questions/11942118/how-do-you-get-gson-to-omit-null-or-empty-objects-and-empty-arrays-and-lists)
    private List<OptionGroupObject> optionGroups;

    private List<OptionObject> initialOptions;
    private ConfirmationDialogObject confirm;
    private Integer maxSelectedItems;

    public MultiStaticSelectElement(String fallback, PlainTextObject placeholder, String actionId, List<OptionObject> options, List<OptionGroupObject> optionGroups, List<OptionObject> initialOptions, ConfirmationDialogObject confirm, Integer maxSelectedItems) {
        this.fallback = fallback;
        this.placeholder = placeholder;
        this.actionId = actionId;
        this.options = options;
        this.optionGroups = optionGroups;
        this.initialOptions = initialOptions;
        this.confirm = confirm;
        this.maxSelectedItems = maxSelectedItems;
    }

    public MultiStaticSelectElement() {
    }

    public static MultiStaticSelectElementBuilder builder() {
        return new MultiStaticSelectElementBuilder();
    }

    public String getType() {
        return this.type;
    }

    public String getFallback() {
        return this.fallback;
    }

    public PlainTextObject getPlaceholder() {
        return this.placeholder;
    }

    public String getActionId() {
        return this.actionId;
    }

    public List<OptionObject> getOptions() {
        return this.options;
    }

    public List<OptionGroupObject> getOptionGroups() {
        return this.optionGroups;
    }

    public List<OptionObject> getInitialOptions() {
        return this.initialOptions;
    }

    public ConfirmationDialogObject getConfirm() {
        return this.confirm;
    }

    public Integer getMaxSelectedItems() {
        return this.maxSelectedItems;
    }

    public void setFallback(String fallback) {
        this.fallback = fallback;
    }

    public void setPlaceholder(PlainTextObject placeholder) {
        this.placeholder = placeholder;
    }

    public void setActionId(String actionId) {
        this.actionId = actionId;
    }

    public void setOptions(List<OptionObject> options) {
        this.options = options;
    }

    public void setOptionGroups(List<OptionGroupObject> optionGroups) {
        this.optionGroups = optionGroups;
    }

    public void setInitialOptions(List<OptionObject> initialOptions) {
        this.initialOptions = initialOptions;
    }

    public void setConfirm(ConfirmationDialogObject confirm) {
        this.confirm = confirm;
    }

    public void setMaxSelectedItems(Integer maxSelectedItems) {
        this.maxSelectedItems = maxSelectedItems;
    }

    public boolean equals(final Object o) {
        if (o == this) return true;
        if (!(o instanceof MultiStaticSelectElement)) return false;
        final MultiStaticSelectElement other = (MultiStaticSelectElement) o;
        if (!other.canEqual((Object) this)) return false;
        if (!super.equals(o)) return false;
        final Object this$type = this.getType();
        final Object other$type = other.getType();
        if (this$type == null ? other$type != null : !this$type.equals(other$type)) return false;
        final Object this$fallback = this.getFallback();
        final Object other$fallback = other.getFallback();
        if (this$fallback == null ? other$fallback != null : !this$fallback.equals(other$fallback)) return false;
        final Object this$placeholder = this.getPlaceholder();
        final Object other$placeholder = other.getPlaceholder();
        if (this$placeholder == null ? other$placeholder != null : !this$placeholder.equals(other$placeholder))
            return false;
        final Object this$actionId = this.getActionId();
        final Object other$actionId = other.getActionId();
        if (this$actionId == null ? other$actionId != null : !this$actionId.equals(other$actionId)) return false;
        final Object this$options = this.getOptions();
        final Object other$options = other.getOptions();
        if (this$options == null ? other$options != null : !this$options.equals(other$options)) return false;
        final Object this$optionGroups = this.getOptionGroups();
        final Object other$optionGroups = other.getOptionGroups();
        if (this$optionGroups == null ? other$optionGroups != null : !this$optionGroups.equals(other$optionGroups))
            return false;
        final Object this$initialOptions = this.getInitialOptions();
        final Object other$initialOptions = other.getInitialOptions();
        if (this$initialOptions == null ? other$initialOptions != null : !this$initialOptions.equals(other$initialOptions))
            return false;
        final Object this$confirm = this.getConfirm();
        final Object other$confirm = other.getConfirm();
        if (this$confirm == null ? other$confirm != null : !this$confirm.equals(other$confirm)) return false;
        final Object this$maxSelectedItems = this.getMaxSelectedItems();
        final Object other$maxSelectedItems = other.getMaxSelectedItems();
        if (this$maxSelectedItems == null ? other$maxSelectedItems != null : !this$maxSelectedItems.equals(other$maxSelectedItems))
            return false;
        return true;
    }

    protected boolean canEqual(final Object other) {
        return other instanceof MultiStaticSelectElement;
    }

    public int hashCode() {
        final int PRIME = 59;
        int result = super.hashCode();
        final Object $type = this.getType();
        result = result * PRIME + ($type == null ? 43 : $type.hashCode());
        final Object $fallback = this.getFallback();
        result = result * PRIME + ($fallback == null ? 43 : $fallback.hashCode());
        final Object $placeholder = this.getPlaceholder();
        result = result * PRIME + ($placeholder == null ? 43 : $placeholder.hashCode());
        final Object $actionId = this.getActionId();
        result = result * PRIME + ($actionId == null ? 43 : $actionId.hashCode());
        final Object $options = this.getOptions();
        result = result * PRIME + ($options == null ? 43 : $options.hashCode());
        final Object $optionGroups = this.getOptionGroups();
        result = result * PRIME + ($optionGroups == null ? 43 : $optionGroups.hashCode());
        final Object $initialOptions = this.getInitialOptions();
        result = result * PRIME + ($initialOptions == null ? 43 : $initialOptions.hashCode());
        final Object $confirm = this.getConfirm();
        result = result * PRIME + ($confirm == null ? 43 : $confirm.hashCode());
        final Object $maxSelectedItems = this.getMaxSelectedItems();
        result = result * PRIME + ($maxSelectedItems == null ? 43 : $maxSelectedItems.hashCode());
        return result;
    }

    public static class MultiStaticSelectElementBuilder {
        private String fallback;
        private PlainTextObject placeholder;
        private String actionId;
        private List<OptionObject> options;
        private List<OptionGroupObject> optionGroups;
        private List<OptionObject> initialOptions;
        private ConfirmationDialogObject confirm;
        private Integer maxSelectedItems;

        MultiStaticSelectElementBuilder() {
        }

        public MultiStaticSelectElement.MultiStaticSelectElementBuilder fallback(String fallback) {
            this.fallback = fallback;
            return this;
        }

        public MultiStaticSelectElement.MultiStaticSelectElementBuilder placeholder(PlainTextObject placeholder) {
            this.placeholder = placeholder;
            return this;
        }

        public MultiStaticSelectElement.MultiStaticSelectElementBuilder actionId(String actionId) {
            this.actionId = actionId;
            return this;
        }

        public MultiStaticSelectElement.MultiStaticSelectElementBuilder options(List<OptionObject> options) {
            this.options = options;
            return this;
        }

        public MultiStaticSelectElement.MultiStaticSelectElementBuilder optionGroups(List<OptionGroupObject> optionGroups) {
            this.optionGroups = optionGroups;
            return this;
        }

        public MultiStaticSelectElement.MultiStaticSelectElementBuilder initialOptions(List<OptionObject> initialOptions) {
            this.initialOptions = initialOptions;
            return this;
        }

        public MultiStaticSelectElement.MultiStaticSelectElementBuilder confirm(ConfirmationDialogObject confirm) {
            this.confirm = confirm;
            return this;
        }

        public MultiStaticSelectElement.MultiStaticSelectElementBuilder maxSelectedItems(Integer maxSelectedItems) {
            this.maxSelectedItems = maxSelectedItems;
            return this;
        }

        public MultiStaticSelectElement build() {
            return new MultiStaticSelectElement(fallback, placeholder, actionId, options, optionGroups, initialOptions, confirm, maxSelectedItems);
        }

        public String toString() {
            return "MultiStaticSelectElement.MultiStaticSelectElementBuilder(fallback=" + this.fallback + ", placeholder=" + this.placeholder + ", actionId=" + this.actionId + ", options=" + this.options + ", optionGroups=" + this.optionGroups + ", initialOptions=" + this.initialOptions + ", confirm=" + this.confirm + ", maxSelectedItems=" + this.maxSelectedItems + ")";
        }
    }
}

File: core/src/commonMain/kotlin/work/socialhub/kslack/entity/block/element/MultiStaticSelectElement.kt
package work.socialhub.kslack.entity.block.element

import work.socialhub.kslack.api.model.block.composition.ConfirmationDialogObject

/**
 * https://api.slack.com/reference/block-kit/block-elements#multi_select
 */
class MultiStaticSelectElement : BlockElement {
    val type: String = TYPE
    var fallback: String? = null

    private var placeholder: PlainTextObject? = null
    var actionId: String? = null

    // https://github.com/seratch/jslack/pull/103
    // The reason I didn't initialize the Array<> fields is because Slack (sometimes) gives errors
    // when it encounters an empty list in the generated JSON.
    // The proper solution if/when you don't want un-initialized fields is to have a Gson type adapter that skips empty lists
    // (e.g. something like https://stackoverflow.com/questions/11942118/how-do-you-get-gson-to-omit-null-or-empty-objects-and-empty-arrays-and-lists)
    private var options: Array<OptionObject>? = null

    // https://github.com/seratch/jslack/pull/103
    // The reason I didn't initialize the Array<> fields is because Slack (sometimes) gives errors
    // when it encounters an empty list in the generated JSON.
    // The proper solution if/when you don't want un-initialized fields is to have a Gson type adapter that skips empty lists
    // (e.g. something like https://stackoverflow.com/questions/11942118/how-do-you-get-gson-to-omit-null-or-empty-objects-and-empty-arrays-and-lists)
    private var optionGroups: Array<OptionGroupObject>? = null

    private var initialOptions: Array<OptionObject>? = null
    private var confirm: ConfirmationDialogObject? = null
    var maxSelectedItems: Int? = null

    constructor(
        fallback: String?,
        placeholder: PlainTextObject?,
        actionId: String?,
        options: Array<OptionObject>?,
        optionGroups: Array<OptionGroupObject>?,
        initialOptions: Array<OptionObject>?,
        confirm: ConfirmationDialogObject?,
        maxSelectedItems: Int?
    ) {
        this.fallback = fallback
        this.placeholder = placeholder
        this.actionId = actionId
        this.options = options
        this.optionGroups = optionGroups
        this.initialOptions = initialOptions
        this.confirm = confirm
        this.maxSelectedItems = maxSelectedItems
    }

    constructor()

    fun getPlaceholder(): PlainTextObject? {
        return this.placeholder
    }

    fun getOptions(): Array<OptionObject>? {
        return this.options
    }

    fun getOptionGroups(): Array<OptionGroupObject>? {
        return this.optionGroups
    }

    fun getInitialOptions(): Array<OptionObject>? {
        return this.initialOptions
    }

    fun getConfirm(): ConfirmationDialogObject? {
        return this.confirm
    }

    fun setPlaceholder(placeholder: PlainTextObject?) {
        this.placeholder = placeholder
    }

    fun setOptions(options: Array<OptionObject>?) {
        this.options = options
    }

    fun setOptionGroups(optionGroups: Array<OptionGroupObject>?) {
        this.optionGroups = optionGroups
    }

    fun setInitialOptions(initialOptions: Array<OptionObject>?) {
        this.initialOptions = initialOptions
    }

    fun setConfirm(confirm: ConfirmationDialogObject?) {
        this.confirm = confirm
    }

    override fun equals(o: Any): Boolean {
        if (o === this) return true
        if (o !is MultiStaticSelectElement) return false
        val other = o
        if (!other.canEqual(this as Any)) return false
        if (!super.equals(o)) return false
        val `this$type`: Any = this.type
        val `other$type`: Any = other.type
        if (if (`this$type` == null) `other$type` != null else `this$type` != `other$type`) return false
        val `this$fallback`: Any? = this.fallback
        val `other$fallback`: Any? = other.fallback
        if (if (`this$fallback` == null) `other$fallback` != null else `this$fallback` != `other$fallback`) return false
        val `this$placeholder`: Any? = this.getPlaceholder()
        val `other$placeholder`: Any? = other.getPlaceholder()
        if (if (`this$placeholder` == null) `other$placeholder` != null else `this$placeholder` != `other$placeholder`) return false
        val `this$actionId`: Any? = this.actionId
        val `other$actionId`: Any? = other.actionId
        if (if (`this$actionId` == null) `other$actionId` != null else `this$actionId` != `other$actionId`) return false
        val `this$options`: Any? = this.getOptions()
        val `other$options`: Any? = other.getOptions()
        if (if (`this$options` == null) `other$options` != null else `this$options` != `other$options`) return false
        val `this$optionGroups`: Any? = this.getOptionGroups()
        val `other$optionGroups`: Any? = other.getOptionGroups()
        if (if (`this$optionGroups` == null) `other$optionGroups` != null else `this$optionGroups` != `other$optionGroups`) return false
        val `this$initialOptions`: Any? = this.getInitialOptions()
        val `other$initialOptions`: Any? = other.getInitialOptions()
        if (if (`this$initialOptions` == null) `other$initialOptions` != null else `this$initialOptions` != `other$initialOptions`) return false
        val `this$confirm`: Any? = this.getConfirm()
        val `other$confirm`: Any? = other.getConfirm()
        if (if (`this$confirm` == null) `other$confirm` != null else `this$confirm` != `other$confirm`) return false
        val `this$maxSelectedItems`: Any? = this.maxSelectedItems
        val `other$maxSelectedItems`: Any? = other.maxSelectedItems
        if (if (`this$maxSelectedItems` == null) `other$maxSelectedItems` != null else `this$maxSelectedItems` != `other$maxSelectedItems`) return false
        return true
    }

    protected fun canEqual(other: Any?): Boolean {
        return other is MultiStaticSelectElement
    }

    override fun hashCode(): Int {
        val PRIME = 59
        var result = super.hashCode()
        val `$type`: Any = this.type
        result = result * PRIME + (`$type`?.hashCode() ?: 43)
        val `$fallback`: Any? = this.fallback
        result = result * PRIME + (`$fallback`?.hashCode() ?: 43)
        val `$placeholder`: Any? = this.getPlaceholder()
        result = result * PRIME + (`$placeholder`?.hashCode() ?: 43)
        val `$actionId`: Any? = this.actionId
        result = result * PRIME + (`$actionId`?.hashCode() ?: 43)
        val `$options`: Any? = this.getOptions()
        result = result * PRIME + (`$options`?.hashCode() ?: 43)
        val `$optionGroups`: Any? = this.getOptionGroups()
        result = result * PRIME + (`$optionGroups`?.hashCode() ?: 43)
        val `$initialOptions`: Any? = this.getInitialOptions()
        result = result * PRIME + (`$initialOptions`?.hashCode() ?: 43)
        val `$confirm`: Any? = this.getConfirm()
        result = result * PRIME + (`$confirm`?.hashCode() ?: 43)
        val `$maxSelectedItems`: Any? = this.maxSelectedItems
        result = result * PRIME + (`$maxSelectedItems`?.hashCode() ?: 43)
        return result
    }

    class MultiStaticSelectElementBuilder() {
        private var fallback: String? = null
        private var placeholder: PlainTextObject? = null
        private var actionId: String? = null
        private var options: Array<OptionObject>? = null
        private var optionGroups: Array<OptionGroupObject>? = null
        private var initialOptions: Array<OptionObject>? = null
        private var confirm: ConfirmationDialogObject? = null
        private var maxSelectedItems: Int? = null

        fun fallback(fallback: String?): MultiStaticSelectElementBuilder {
            this.fallback = fallback
            return this
        }

        fun placeholder(placeholder: PlainTextObject?): MultiStaticSelectElementBuilder {
            this.placeholder = placeholder
            return this
        }

        fun actionId(actionId: String?): MultiStaticSelectElementBuilder {
            this.actionId = actionId
            return this
        }

        fun options(options: Array<OptionObject>?): MultiStaticSelectElementBuilder {
            this.options = options
            return this
        }

        fun optionGroups(optionGroups: Array<OptionGroupObject>?): MultiStaticSelectElementBuilder {
            this.optionGroups = optionGroups
            return this
        }

        fun initialOptions(initialOptions: Array<OptionObject>?): MultiStaticSelectElementBuilder {
            this.initialOptions = initialOptions
            return this
        }

        fun confirm(confirm: ConfirmationDialogObject?): MultiStaticSelectElementBuilder {
            this.confirm = confirm
            return this
        }

        fun maxSelectedItems(maxSelectedItems: Int?): MultiStaticSelectElementBuilder {
            this.maxSelectedItems = maxSelectedItems
            return this
        }

        fun build(): MultiStaticSelectElement {
            return MultiStaticSelectElement(
                fallback,
                placeholder,
                actionId,
                options,
                optionGroups,
                initialOptions,
                confirm,
                maxSelectedItems
            )
        }

        override fun toString(): String {
            return "MultiStaticSelectElement.MultiStaticSelectElementBuilder(fallback=" + this.fallback + ", placeholder=" + this.placeholder + ", actionId=" + this.actionId + ", options=" + this.options + ", optionGroups=" + this.optionGroups + ", initialOptions=" + this.initialOptions + ", confirm=" + this.confirm + ", maxSelectedItems=" + this.maxSelectedItems + ")"
        }
    }

    companion object {
        const val TYPE: String = "multi_static_select"
        fun builder(): MultiStaticSelectElementBuilder {
            return MultiStaticSelectElementBuilder()
        }
    }
}
